File: internal/cipher/cipher.go
package cipher

import "encoding/binary"

const DefaultKey = "instahost"
const ChecksumBytes = 4

var crc32Table [256]uint32

func init() {
	for i := 0; i < 256; i++ {
		c := uint32(i)
		for j := 0; j < 8; j++ {
			if c&1 != 0 {
				c = 0xedb88320 ^ (c >> 1)
			} else {
				c >>= 1
			}
		}
		crc32Table[i] = c
	}
}

func CRC32(data []byte) uint32 {
	crc := uint32(0xffffffff)
	for _, b := range data {
		crc = crc32Table[(crc^uint32(b))&0xff] ^ (crc >> 8)
	}
	return (crc ^ 0xffffffff)
}

func SealWithChecksum(data []byte) []byte {
	out := make([]byte, len(data)+ChecksumBytes)
	copy(out, data)
	binary.BigEndian.PutUint32(out[len(data):], CRC32(data))
	return out
}

func OpenWithChecksum(data []byte) ([]byte, bool) {
	if len(data) <= ChecksumBytes {
		return nil, false
	}
	payload := data[:len(data)-ChecksumBytes]
	stored := binary.BigEndian.Uint32(data[len(data)-ChecksumBytes:])
	if stored != CRC32(payload) {
		return nil, false
	}
	return payload, true
}

func XorBytes(data []byte, key string) []byte {
	if key == "" {
		key = DefaultKey
	}
	keyBytes := []byte(key)
	out := make([]byte, len(data))
	for i, b := range data {
		out[i] = b ^ keyBytes[i%len(keyBytes)]
	}
	return out
}

File: lib/cipher.js
export const DEFAULT_KEY = "instahost";
export const CHECKSUM_BYTES = 4;

const CRC32_TABLE = new Uint32Array(256);
for (let i = 0; i < 256; i++) {
  let c = i;
  for (let j = 0; j < 8; j++) {
    c = (c & 1) ? (0xedb88320 ^ (c >>> 1)) : (c >>> 1);
  }
  CRC32_TABLE[i] = c;
}

export function crc32(data) {
  const input = Buffer.isBuffer(data) ? data : Buffer.from(data);
  let crc = 0xffffffff;

  for (let i = 0; i < input.length; i++) {
    crc = CRC32_TABLE[(crc ^ input[i]) & 0xff] ^ (crc >>> 8);
  }

  return (crc ^ 0xffffffff) >>> 0;
}

export function sealWithChecksum(data) {
  const input = Buffer.isBuffer(data) ? data : Buffer.from(data);
  const out = Buffer.alloc(input.length + CHECKSUM_BYTES);
  input.copy(out);
  out.writeUInt32BE(crc32(input), input.length);
  return out;
}

export function openWithChecksum(data) {
  const input = Buffer.isBuffer(data) ? data : Buffer.from(data);
  if (input.length <= CHECKSUM_BYTES) {
    return null;
  }

  const payload = input.subarray(0, input.length - CHECKSUM_BYTES);
  const stored = input.readUInt32BE(input.length - CHECKSUM_BYTES);
  if (stored !== crc32(payload)) {
    return null;
  }

  return payload;
}

export function xorBytes(data, key = DEFAULT_KEY) {
  const keyBytes = Buffer.from(key, "utf8");
  const input = Buffer.isBuffer(data) ? data : Buffer.from(data);
  const out = Buffer.alloc(input.length);

  for (let i = 0; i < input.length; i++) {
    out[i] = input[i] ^ keyBytes[i % keyBytes.length];
  }

  return out;
}
